File: src/python/image_io.py
from __future__ import annotations

import os
from pathlib import Path

import numpy as np
from PIL import Image


def load_color_image(path: str | os.PathLike[str]) -> Image.Image:
    image = Image.open(path).convert("RGB")
    if image.width == 0 or image.height == 0:
        raise ValueError(f"empty image: {path}")
    return image


def ensure_directory(directory: str | os.PathLike[str]) -> None:
    Path(directory).mkdir(parents=True, exist_ok=True)


def make_output_path(
    input_path: str | os.PathLike[str],
    output_dir: str | os.PathLike[str],
    clusters: int,
    quality: int,
) -> Path:
    stem = Path(input_path).stem
    filename = f"{stem}_k{clusters}_q{quality}.jpg"
    return Path(output_dir) / filename


def save_jpeg(path: str | os.PathLike[str], image: Image.Image, quality: int) -> None:
    image.save(path, format="JPEG", quality=quality, optimize=True)


def image_to_array(image: Image.Image) -> np.ndarray:
    return np.asarray(image, dtype=np.uint8)


def array_to_image(array: np.ndarray) -> Image.Image:
    if array.dtype != np.uint8:
        raise ValueError("array_to_image expects dtype=uint8")
    if array.ndim != 3 or array.shape[2] != 3:
        raise ValueError("array_to_image expects shape (H, W, 3)")
    return Image.fromarray(array, mode="RGB")


def flatten_image(array: np.ndarray) -> np.ndarray:
    if array.ndim != 3 or array.shape[2] != 3:
        raise ValueError("flatten_image expects shape (H, W, 3)")
    return array.reshape(-1, 3).astype(np.float32)


def unflatten_image(samples: np.ndarray, rows: int, cols: int) -> np.ndarray:
    if samples.ndim != 2 or samples.shape[1] != 3:
        raise ValueError("unflatten_image expects shape (N, 3)")
    if samples.shape[0] != rows * cols:
        raise ValueError("sample count does not match rows * cols")
    reshaped = samples.reshape(rows, cols, 3)
    return np.clip(reshaped, 0, 255).astype(np.uint8)

File: src/cpp/src/image_io.cpp
#include "image_io.hpp"

#include <opencv2/imgcodecs.hpp>

#include <filesystem>
#include <sstream>
#include <stdexcept>
#include <vector>

namespace fs = std::filesystem;

namespace imgc::io {

cv::Mat load_color_image(const std::string& path) {
    cv::Mat image = cv::imread(path, cv::IMREAD_COLOR);
    if (image.empty()) {
        throw std::runtime_error("failed to read image: " + path);
    }
    return image;
}

void ensure_directory(const std::string& directory) {
    std::error_code ec;
    fs::create_directories(directory, ec);
    if (ec) {
        throw std::runtime_error("failed to create directory: " + directory);
    }
}

std::string make_output_path(const std::string& input_path, const std::string& output_dir, int clusters, int quality) {
    fs::path input(input_path);
    std::ostringstream name;
    name << input.stem().string() << "_k" << clusters << "_q" << quality << ".jpg";
    return (fs::path(output_dir) / name.str()).string();
}

void save_jpeg(const std::string& path, const cv::Mat& image, int quality) {
    std::vector<int> params{cv::IMWRITE_JPEG_QUALITY, quality};
    if (!cv::imwrite(path, image, params)) {
        throw std::runtime_error("failed to write image: " + path);
    }
}

}  // namespace imgc::io

